src/analysis/souffle/examples/simple_may_will/simple_may_will_pass.authlogic

```diff
@@ -0,0 +1,31 @@
+//-----------------------------------------------------------------------------
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//-------------------------------------------------------------------------------
+
+"EndUser" says {
+    ownsTag("EndUser", "userSelection").
+    "P1" canSay hasTag(PathX, "userSelection") :- isAccessPath(PathX).
+    may("P2", "doLocalComputation", "userSelection").
```

Comment:
Isn't something missing here? Don't we want to say that "P2" is allowed to do `"localComputation"` on the data `"EndUser"` owns as long as it has the tag `"userSelection"`?

Reply:
It's the other way: the taint analysis tells us if data _might_ be influenced by user selection data, not that it definitely was.  The idea behind the check is to check for consent for usage whenever data may have influenced something.

Ah although, the example is probably confusing because "userSelection" is meant to be a secret, but that doesn't seem intuitively like a secret. I changed it to "rawVideo" so that the example makes more sense conceptually.

I left this open to see if it actually addresses your question/comment.

Reply:
I probably did not word my comment well.  In this example, we are hard coding "P2" in the authorization logic. If so, doesn't that require "EndUser" to be aware of all particles in the system? Ideally, the "EndUser" will only be aware of the ingestion particle at the edge of the data flow graph. Therefore, "EndUser" will allow the ingestion particle to add tags to the data that capture the data usage restrictions. In this case, we might want to say that the tag `"userSelection"` (or `"rawVideo"`) would mean only "doLocalComputation" is allowed. As the data flows through the system, the policy engine should make sure that the tags on the data are compatible with the operations being performed. I am not sure how that would be modeled or captured.

Reply:
> In this example, we are hard coding "P2" in the authorization logic. If so, doesn't that require "EndUser" to be aware of all particles in the system? 

This is true in this simple test, but that's not true in general. In a real example, "EndUser" would delegate to a party it trusts to apply the tags properly, say "ThirdPartyPrivacyOrg". Then "ThirdPartyPrivacyOrg" would audit the interfaces of the actual application (in other words the ingestion particles) and apply the tags to the particles (by delegating to the particles)**. Concretely this would look like:

```
"EndUser" says "ThirdPartyPrivacyOrg" canSay hasTag(pathX, "rawVideo") :- isPath(pathX).
"ThirdPartyPrivacyOrg" says {
     "P2" canSay hasTag(pathX, "rawVideo") :- isPath(pathX).
     "P3" canSay hasTag(pathX, "rawVideo") :- isPath(pathX).
     // ..etc
}
```
As a result, the user's policy only needs to talk about the abstraction of the data and top-level organzations. The third party privacy org's policy talks about the actual particles because it audits the application. More layers of delegation could also be introduced as needed (e.g. user delegates to SomeCompnay, and SomeCompany delegates to PrivacySubOrg). 

Although this one is simple because it's a test, these two are meant to be more fully fleshed out examples with a delegation chain that, like the one described in this thread, does not require the user to know about the application: [1](https://github.com/google-research/raksha/tree/main/rust/tools/authorization-logic/raksha_examples), [2](https://github.com/google-research/raksha/pull/197).


> As the data flows through the system, the policy engine should make sure that the tags on the data are compatible with the operations being performed. I am not sure how that would be modeled or captured.

This PR is meant to implement exactly this check. Just like you said, the tags are added to the ingestion points. Then the taint analysis (in `taint.dl`) propagates these tags. The check implemented here (mainly in `may_will.dl` and `datalog_facts.h`) is meant to make sure the operations done on data tainted with the tags matches what the data owners say is allowed. The auditors of the application behavior spec write what particles will do and this populates `saysWill(...)`. The rule that defines `disallowedUsage` finds the cases where there is a mismatch, which happens when a particle `will` do something with a handle (saysWill on the RHS), that has been tainted with a tag (mayHaveTag on the RHS), and the owner of the tag does not allow it (!permittedUsage) on the RHS.

**Delegation  with canSay does already work like this -- if X delegates to Y, Y can delegate to Z. If we want add another form of delegation that prevents re-delegation, SecPal also has a design for this we can just pick up.

Reply:
Sounds good.  My other question is if using tags in the  `will` and `may` abstractions are sufficient for capturing all the desired operations in a particle. In any case, we can solve that later and let us go with this for the time being.

Reply:
The goal here is to check prescriptive policies about data usage. 

BTW this idea is mostly made up. It's based on the check done in [This policy language paper](https://www.microsoft.com/en-us/research/publication/s4p-a-generic-language-for-specifying-privacy-preferences-and-policies/) (that you sent out), except that in our implementation, the check involves a static (taint) analysis, whereas in the paper, this may/will check happens when a client talks to a remote service.